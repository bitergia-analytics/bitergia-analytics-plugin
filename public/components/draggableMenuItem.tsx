/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import {
  EuiDraggable,
  EuiFlexItem,
  EuiFlexGroup,
  EuiPanel,
  EuiButtonIcon,
  EuiText,
} from '@elastic/eui';

export const DraggableMenuItem = ({ name, index, parentIndex, removeItem }) => {
  return (
    <EuiDraggable
      index={index}
      draggableId={`parent_${parentIndex}_${index}`}
      spacing="s"
      isRemovable={true}
    >
      <EuiPanel
        paddingSize="s"
        grow={false}
        style={{
          border: '0',
          boxShadow: 'none',
          background: 'rgba(110, 169, 264, 0.05)',
        }}
      >
        <EuiFlexGroup justifyContent="spaceBetween" gutterSize="xs">
          <EuiFlexItem>
            <EuiText size="s">{name}</EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButtonIcon
              iconType="cross"
              aria-label={`Remove ${name}`}
              onClick={() => removeItem(parentIndex, index)}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiPanel>
    </EuiDraggable>
  );
};
