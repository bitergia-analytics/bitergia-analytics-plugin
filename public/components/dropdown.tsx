import React, { useState, useEffect } from 'react';
import {
  EuiHeaderLink,
  EuiPopover,
  EuiPopoverTitle,
  EuiListGroup
} from '@elastic/eui';

const Dropdown = ({ item }) => {
  const [isOpen, setIsOpen] = useState(false);
  const dashboards = getDashboardLinks(item.dashboards);

  function getDashboardLinks(dashboards) {
    return dashboards.map(dashboard => {
      return {
        label: dashboard.name,
        href: `dashboards#/view/${dashboard.panel_id}`
      }
    });
  }

  function handleClick() {
    setIsOpen(!isOpen);
  }

  function closePopover() {
    setIsOpen(false);
  }

  return (
    <EuiPopover
      button={
        <EuiHeaderLink
          iconType='arrowDown'
          iconSide='right'
          onClick={handleClick}
          isActive={item.isActive}
        >
          {item.name}
        </EuiHeaderLink>
      }
      anchorPosition="downCenter"
      isOpen={isOpen}
      closePopover={closePopover}
      panelPaddingSize="m"
      panelClassName="custom-panel"
    >
      <EuiPopoverTitle paddingsize="l">
        {item.name}
      </EuiPopoverTitle>
      <EuiListGroup
        listItems={dashboards}
        flush={true}
        size="s"
        gutterSize="none"
        onClick={closePopover}
      />
    </EuiPopover>
  );
}

export default Dropdown;
