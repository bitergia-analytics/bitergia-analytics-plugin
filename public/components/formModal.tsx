/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useEffect, useState } from 'react';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiFieldText,
  EuiForm,
  EuiFormRow,
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiRadioGroup,
  EuiOverlayMask,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiTitle,
  EuiButtonIcon,
  EuiComboBox,
} from '@elastic/eui';

export const FormModal = ({ showModal, saveItem, item, http }) => {
  const initialValues = {
    isModalVisible: false,
    isInvalid: false,
    errors: [],
    name: '',
    type: 'entry',
    dashboardId: {
      label: '',
      dashboard_id: '',
    },
    dashboards: [
      {
        name: '',
        type: 'entry',
        label: '',
        dashboard_id: '',
      },
    ],
  };
  const [isModalVisible, setIsModalVisible] = useState(
    initialValues.isModalVisible
  );
  const [isInvalid, setIsInvalid] = useState(initialValues.isInvalid);
  const [errors, setErrors] = useState(initialValues.errors);
  const [name, setName] = useState(initialValues.name);
  const [type, setType] = useState(initialValues.type);
  const [dashboardId, setDashboardId] = useState(initialValues.dashboardId);
  const [dashboards, setDashboards] = useState(initialValues.dashboards);
  const [options, setOptions] = useState([]);

  useEffect(() => {
    function findOption(id) {
      if (!id) {
        return;
      }
      const found = options.find((option) => option.dashboard_id === id);
      return found ? found : createOption(id, options);
    }

    setIsModalVisible(showModal || initialValues.showModal);
    setName(item?.name || initialValues.name);
    setType(item?.type || initialValues.type);
    setDashboardId(findOption(item?.dashboard_id) || initialValues.dashboardId);
    setDashboards(initialValues.dashboards);

    if (item?.dashboards) {
      const copy = [...item.dashboards];
      const dashboardsData = copy.map((dashboard) => {
        const optionData = findOption(dashboard.dashboard_id);
        return { ...optionData, ...dashboard };
      });
      setDashboards(dashboardsData);
    }
  }, [showModal, item]);

  useEffect(() => {
    async function fetchDashboards() {
      try {
        const response = await http.put('/api/dashboards/search');

        const result = response.hits.map((hit) => {
          return {
            label: hit._source.dashboard.title,
            dashboard_id: hit._id.replace('dashboard:', ''),
          };
        });

        setOptions(result);
      } catch (error) {
        console.log(error);
      }
    }

    fetchDashboards();
  }, []);

  const addDashboard = () => {
    setDashboards([
      ...dashboards,
      {
        name: '',
        type: 'entry',
        dashboard_id: '',
        label: '',
      },
    ]);
  };

  const removeDashboard = (index) => {
    const copy = [...dashboards];
    copy.splice(index, 1);
    setDashboards(copy);
  };

  const createOption = (searchValue = '', flattenedOptions = []) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const newOption = {
      label: searchValue,
      dashboard_id: searchValue,
    };

    if (
      !flattenedOptions.find(
        (option) =>
          option.dashboard_id.trim().toLowerCase() === normalizedSearchValue
      )
    ) {
      setOptions([...options, newOption]);
    }

    return newOption;
  };

  const setDashboardData = (index, value) => {
    const copy = [...dashboards];
    Object.assign(copy[index], value);
    setDashboards(copy);
  };

  const closeModal = () => {
    setIsInvalid(false);
    setErrors([]);
    setIsModalVisible(false);
    setName(initialValues.name);
    setType(initialValues.type);
    setDashboardId(initialValues.dashboardId);
    setDashboards(initialValues.dashboards);
  };

  const onSave = () => {
    setIsInvalid(false);
    const messages = [];
    if (!name) {
      messages.push('Name is required');
    }
    if (type === 'entry' && !dashboardId.dashboard_id) {
      messages.push('Dashboard ID is required');
    }
    if (type === 'menu' && dashboards.some((d) => !d.name || !d.dashboard_id)) {
      messages.push('Missing menu item data');
    }
    if (messages.length > 0) {
      setIsInvalid(true);
      setErrors(messages);
      return;
    }

    const itemData = { name, type };

    if (type === 'entry') {
      itemData.dashboard_id = dashboardId.dashboard_id;
    } else {
      itemData.dashboards = dashboards.map((dashboard) => {
        return {
          dashboard_id: dashboard.dashboard_id,
          name: dashboard.name,
          type: 'entry',
        };
      });
    }

    saveItem(itemData);
    closeModal();
  };

  const entryForm = (
    <>
      <EuiSpacer size="s" />
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFormRow label="Name" id="name">
            <EuiFieldText
              name="name"
              id="name"
              value={name}
              onChange={(e) => setName(e.target.value)}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Dashboard ID or URL" id="dashboard-id">
            <EuiComboBox
              options={options}
              selectedOptions={[dashboardId]}
              singleSelection={{ asPlainText: true }}
              onChange={(e) =>
                setDashboardId(e[0] || initialValues.dashboardId)
              }
              onCreateOption={(e) => setDashboardId(createOption(e))}
              id="dashboard-id"
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </>
  );

  const dashboardsForm = dashboards.map((dashboard, index) => (
    <div key={index}>
      <EuiFlexGroup alignItems="center">
        <EuiFlexItem>
          <EuiFormRow label="Name">
            <EuiFieldText
              name="name"
              value={dashboard.name}
              onChange={(e) =>
                setDashboardData(index, { name: e.target.value })
              }
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Dashboard ID or URL">
            <EuiComboBox
              options={options}
              selectedOptions={[dashboards[index]]}
              singleSelection={{ asPlainText: true }}
              onChange={(e) =>
                setDashboardData(
                  index,
                  e[0] || {
                    dashboard_id: '',
                    label: '',
                  }
                )
              }
              onCreateOption={(e) => setDashboardData(index, createOption(e))}
              style={{ width: '350px' }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiFormRow hasEmptyLabelSpace>
            <EuiButtonIcon
              size="s"
              onClick={() => {
                removeDashboard(index);
              }}
              iconType="cross"
              aria-label="Remove item"
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  ));

  const menuForm = (
    <>
      <EuiFormRow label="Menu name">
        <EuiFieldText
          name="name"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </EuiFormRow>
      <div>
        <EuiSpacer size="m" />
        <EuiTitle size="xxs">
          <h2>Menu items</h2>
        </EuiTitle>
        <EuiSpacer size="m" />
        {dashboardsForm}
        <EuiButtonEmpty onClick={addDashboard}>Add more items</EuiButtonEmpty>
      </div>
    </>
  );

  return (
    <div>
      {isModalVisible && (
        <EuiOverlayMask>
          <EuiModal onClose={closeModal} style={{ width: 800 }}>
            <EuiModalHeader>
              <EuiModalHeaderTitle>
                <h1>{item?.name ? 'Edit item' : 'Add item'}</h1>
              </EuiModalHeaderTitle>
            </EuiModalHeader>

            <EuiModalBody>
              <EuiForm component="form" isInvalid={isInvalid} error={errors}>
                <EuiFormRow label="Type" id="type">
                  <EuiRadioGroup
                    options={[
                      {
                        id: 'entry',
                        label: 'Link',
                      },
                      {
                        id: 'menu',
                        label: 'Dropdown menu',
                      },
                    ]}
                    idSelected={type}
                    onChange={(id) => setType(id)}
                    name="type"
                    id="type"
                  />
                </EuiFormRow>

                {type === 'entry' ? entryForm : menuForm}
              </EuiForm>
            </EuiModalBody>

            <EuiModalFooter>
              <EuiButtonEmpty onClick={closeModal}>Cancel</EuiButtonEmpty>
              <EuiButton type="submit" onClick={onSave} fill>
                Save
              </EuiButton>
            </EuiModalFooter>
          </EuiModal>
        </EuiOverlayMask>
      )}
    </div>
  );
};
