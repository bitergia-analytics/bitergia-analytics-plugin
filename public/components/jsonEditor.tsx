/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState, useEffect } from 'react';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiCodeEditor,
  EuiFlexGroup,
  EuiFlexItem,
  EuiForm,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';

export const JsonEditor = ({ http, renderToast, value }) => {
  const [jsonValue, setJsonValue] = useState(JSON.stringify({}));
  const [isInvalid, setIsInvalid] = useState(false);
  const [isChanged, setIsChanged] = useState(false);
  const [errors, setErrors] = useState([]);

  const placeholderJson = {
    metadashboard: [
      {
        name: 'Panel name',
        type: 'entry',
        dashboard_id: 'id',
      },
    ],
  };

  const metadashboard = value || placeholderJson;

  const setInitialValue = () => {
    setJsonValue(JSON.stringify(metadashboard, null, 2));
    setErrors([]);
  };

  const onCodeEditorChange = (value) => {
    try {
      JSON.parse(value);
      setErrors([]);
    } catch (error) {
      setErrors(['Invalid JSON syntax']);
    }
    setJsonValue(value);
  };

  const onSave = async () => {
    try {
      const response = await http.put('/api/dashboards/metadashboard/edit', {
        body: jsonValue,
      });
      renderToast();
    } catch (error) {
      const message = error.body?.message || error.toString();
      setErrors([message]);
      renderToast(error);
    }
  };

  useEffect(() => {
    setInitialValue();
  }, [value]);

  useEffect(() => {
    const isInvalid = errors.length > 0;
    setIsInvalid(isInvalid);
  }, [errors]);

  useEffect(() => {
    const oldValue = JSON.stringify(metadashboard, null, 2);
    setIsChanged(oldValue !== jsonValue);
  }, [jsonValue]);

  return (
    <EuiForm>
      <EuiFormRow label="JSON" isInvalid={isInvalid} error={errors} fullWidth>
        <EuiCodeEditor
          mode="hjson"
          value={jsonValue}
          onChange={onCodeEditorChange}
          width="100%"
          height="auto"
          minLines={6}
          maxLines={20}
          showGutter={false}
          fontSize={14}
          setOptions={{
            showLineNumbers: false,
            tabSize: 2,
            wrap: true,
          }}
        />
      </EuiFormRow>
      <EuiSpacer size="m" />
      <EuiFlexGroup>
        <EuiFlexItem grow={false}>
          <EuiButton
            onClick={onSave}
            disabled={isInvalid || !isChanged}
            iconType="check"
            fill
          >
            Save changes
          </EuiButton>
        </EuiFlexItem>
        {isChanged && (
          <EuiFlexItem grow={false}>
            <EuiButtonEmpty
              iconType="cross"
              flush="left"
              onClick={setInitialValue}
            >
              Cancel changes
            </EuiButtonEmpty>
          </EuiFlexItem>
        )}
      </EuiFlexGroup>
    </EuiForm>
  );
};
