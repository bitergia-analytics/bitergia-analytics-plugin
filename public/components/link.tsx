import React from 'react';
import { EuiHeaderLink } from '@elastic/eui';

const Link = ({ item }) => {
  return(
    <EuiHeaderLink
     href={`dashboards#/view/${item.panel_id}`}
     isActive={item.isActive}
     >
      { item.name }
    </EuiHeaderLink>
  );
};

export default Link;
