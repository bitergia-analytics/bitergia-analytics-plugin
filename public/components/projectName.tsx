/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import {
  EuiBadge,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHeaderLink,
} from '@elastic/eui';

export const ProjectName = ({ name, tenant, badgeColor, baseURL }) => {
  return (
    <EuiFlexGroup
      grow="false"
      responsive={false}
      alignItems="center"
      gutterSize="xs"
    >
      <EuiFlexItem grow={false}>
        <EuiHeaderLink
          href={`${baseURL}#/view/Overview`}
          className="project-link"
          flush="left"
        >
          {name}
        </EuiHeaderLink>
      </EuiFlexItem>
      {tenant && (
        <EuiFlexItem grow={false}>
          <EuiBadge color={badgeColor}>{tenant}</EuiBadge>
        </EuiFlexItem>
      )}
    </EuiFlexGroup>
  );
};
