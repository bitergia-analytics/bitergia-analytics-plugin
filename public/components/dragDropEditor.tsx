/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState, useEffect } from 'react';
import {
  EuiDragDropContext,
  EuiDraggable,
  EuiDroppable,
  EuiFlexItem,
  EuiFlexGroup,
  EuiPanel,
  EuiButtonIcon,
  EuiIcon,
  EuiText,
  EuiButton,
  EuiSpacer,
} from '@elastic/eui';
import { FormModal } from './formModal';
import { DraggableMenuItem } from './draggableMenuItem';

export const DragDropEditor = ({ value, http, renderToast }) => {
  const [metadashboard, setMetadashboard] = useState(value || []);
  const [showModal, setShowModal] = useState(false);
  const [editableItem, setEditableItem] = useState();

  const openModal = (item, index = metadashboard.length + 1) => {
    setShowModal(true);
    setEditableItem({ ...item, index });
  };

  const saveItem = (item) => {
    const copy = [...metadashboard];
    copy.splice(editableItem.index, 1, item);
    setMetadashboard(copy);
  };

  const removeItem = (parentIndex, index) => {
    const copy = [...metadashboard];
    if (index !== undefined) {
      copy[parentIndex].dashboards.splice(index, 1);
    } else {
      copy.splice(parentIndex, 1);
    }
    setMetadashboard(copy);
  };

  const onSave = async () => {
    try {
      const body = JSON.stringify({ metadashboard });
      const response = await http.put('/api/dashboards/metadashboard/edit', {
        body,
      });
      renderToast();
    } catch (error) {
      const message = error.body?.message || error.toString();
      renderToast(error);
    }
  };

  const onDragEnd = ({ source, destination }) => {
    const copy = [...metadashboard];
    if (source && destination) {
      if (source.droppableId === 'parent') {
        const item = copy.splice(source.index, 1);
        copy.splice(destination.index, 0, item[0]);
      } else {
        const item = copy[source.droppableId].dashboards.splice(
          source.index,
          1
        );
        copy[destination.droppableId].dashboards.splice(
          destination.index,
          0,
          item[0]
        );
      }
      setMetadashboard(copy);
    }
  };

  return (
    <>
      <FormModal
        saveItem={saveItem}
        showModal={showModal}
        item={editableItem}
        http={http}
      />

      <EuiFlexGroup alignItems="center">
        <EuiFlexItem>
          <EuiText size="s">Drag and drop items to rearrange them</EuiText>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton onClick={() => openModal({})}>Add item</EuiButton>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton onClick={onSave} fill>
            Save changes
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>

      <EuiSpacer />

      <EuiDragDropContext onDragEnd={onDragEnd}>
        <EuiDroppable
          direction="horizontal"
          droppableId="parent"
          type="macro"
          style={{
            display: 'flex',
            overflowY: 'auto',
          }}
        >
          {metadashboard.map((dashboard, index) => (
            <EuiDraggable
              key={index}
              index={index}
              draggableId={`parent_${index}`}
              spacing="s"
              isRemovable={true}
            >
              {(provided, state) => (
                <EuiPanel paddingSize="s" grow={false}>
                  <EuiFlexGroup justifyContent="spaceBetween">
                    <EuiFlexItem grow={false}>
                      <EuiText size="s">{dashboard.name}</EuiText>
                    </EuiFlexItem>
                    <EuiFlexItem grow={false}>
                      <EuiFlexGroup gutterSize="xs">
                        <EuiFlexItem>
                          <EuiButtonIcon
                            iconType="pencil"
                            iconSize="s"
                            aria-label="Edit"
                            onClick={() => openModal(dashboard, index)}
                          />
                        </EuiFlexItem>
                        <EuiFlexItem>
                          <EuiButtonIcon
                            iconType="cross"
                            aria-label="Remove"
                            onClick={() => removeItem(index)}
                          />
                        </EuiFlexItem>
                      </EuiFlexGroup>
                    </EuiFlexItem>
                  </EuiFlexGroup>
                  {metadashboard[index].dashboards && (
                    <EuiDroppable droppableId={index.toString()} spacing="s">
                      {metadashboard[index].dashboards.map(({ name }, idx) => (
                        <DraggableMenuItem
                          name={name}
                          key={idx}
                          index={idx}
                          parentIndex={index}
                          removeItem={removeItem}
                        />
                      ))}
                    </EuiDroppable>
                  )}
                </EuiPanel>
              )}
            </EuiDraggable>
          ))}
        </EuiDroppable>
      </EuiDragDropContext>
    </>
  );
};
