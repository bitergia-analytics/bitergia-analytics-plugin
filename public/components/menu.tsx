/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState, useEffect } from 'react';
import { EuiHeaderLinks } from '@elastic/eui';
import { Link } from './link.tsx';
import { Dropdown } from './dropdown.tsx';

export const Menu = (props) => {
  const [metadashboard, setMetadashboard] = useState(props.metadashboard);

  function setActiveLink() {
    let updated = metadashboard;

    if (props.history.location.hash.includes('#/view/')) {
      const currentDashboard = props.history.location.hash
        .split('#/view/')[1]
        .split('?')[0];

      updated = metadashboard.map((dashboard) => {
        if (dashboard.type === 'entry') {
          return Object.assign(dashboard, {
            isActive: currentDashboard === dashboard.panel_id,
          });
        }
        return Object.assign(dashboard, {
          isActive: dashboard.dashboards.some(
            (d) => currentDashboard === d.panel_id
          ),
        });
      });
    } else {
      updated = metadashboard.map((dashboard) =>
        Object.assign(dashboard, { isActive: false })
      );
    }

    setMetadashboard(updated);
  }

  useEffect(() => {
    // Set active link on first load
    setActiveLink();

    // Add class to parent container
    const menu = document.querySelector('.bitergia-menu');
    const parent = menu.closest('.euiHeaderSectionItem').parentElement;
    parent.classList.add('bitergia-menu-parent');
  }, []);

  useEffect(() => {
    // Listen to route changes
    const unlisten = props.history.listen(setActiveLink);
    return unlisten;
  }, []);

  return (
    <EuiHeaderLinks gutterSize="xs" className="bitergia-menu">
      {metadashboard.map((link, index) =>
        link.type === 'entry' ? (
          <Link item={link} key={index} />
        ) : (
          <Dropdown item={link} key={index} />
        )
      )}
    </EuiHeaderLinks>
  );
};
