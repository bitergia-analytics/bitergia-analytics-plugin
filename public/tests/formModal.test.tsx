/*
 * Copyright 2021-2022 Bitergia
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import {
  render,
  screen,
  act,
  waitFor,
  fireEvent,
} from '@testing-library/react';
import { FormModal } from '../components/formModal';
import { coreServicesMock } from '../../test/mocks';

describe('<FormModal />', () => {
  it('Renders the component', async () => {
    // Needs to be wrapped in act() due to asynchronous state update on load
    await act(async () => {
      render(
        <FormModal
          showModal={true}
          item={{ name: 'name' }}
          http={coreServicesMock.http}
        />
      );

      // Wait for state change and component render
      await waitFor(() => {
        expect(document.querySelector('.euiModal')).toBeVisible();
        expect(document.querySelector('.euiModal')).toMatchSnapshot();
      });
    });
  });

  it('Saves data', async () => {
    const saveItem = jest.fn();
    const dashboardID = '12345-abc';

    await act(async () => {
      render(
        <FormModal
          showModal={true}
          item={{ name: 'Test name' }}
          http={coreServicesMock.http}
          saveItem={saveItem}
        />
      );

      await waitFor(() => {
        expect(document.querySelector('.euiModal')).toBeVisible();
      });
    });

    // Fill in dashboard ID field
    const dashboardInput = screen.getByLabelText('Dashboard ID or URL');
    fireEvent.change(dashboardInput, {
      target: { value: [dashboardID] },
    });
    fireEvent.keyDown(dashboardInput, { key: 'Enter' });
    fireEvent.click(screen.getByText('Save'));

    expect(saveItem).toHaveBeenCalledWith({
      dashboard_id: dashboardID,
      name: 'Test name',
      type: 'entry',
    });
  });

  it('Renders errors', async () => {
    const saveItem = jest.fn();
    await act(async () => {
      render(
        <FormModal
          showModal={true}
          http={coreServicesMock.http}
          saveItem={saveItem}
        />
      );

      await waitFor(() => {
        expect(document.querySelector('.euiModal')).toBeVisible();
      });
    });

    fireEvent.click(screen.getByText('Save'));

    expect(saveItem).not.toBeCalled();
    expect(
      screen.getByText('Please address the highlighted errors.')
    ).toBeVisible();
    expect(screen.getByText('Name is required')).toBeVisible();
    expect(screen.getByText('Dashboard ID is required')).toBeVisible();
  });
});
