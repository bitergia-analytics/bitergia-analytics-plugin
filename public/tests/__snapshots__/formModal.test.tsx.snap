// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormModal /> Renders the component 1`] = `
<div
  class="euiModal euiModal--maxWidth-default"
  style="width: 800px;"
  tabindex="0"
>
  <button
    aria-label="Closes this modal window"
    class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
    type="button"
  >
    <div
      aria-hidden="true"
      class="euiButtonIcon__icon"
      data-euiicon-type="cross"
    />
  </button>
  <div
    class="euiModal__flex"
  >
    <div
      class="euiModalHeader"
    >
      <div
        class="euiModalHeader__title"
      >
        <h1>
          Edit item
        </h1>
      </div>
    </div>
    <div
      class="euiModalBody"
    >
      <div
        class="euiModalBody__overflow"
      >
        <form
          class="euiForm"
        >
          <div
            class="euiFormRow"
            id="type-row"
          >
            <div
              class="euiFormRow__labelWrapper"
            >
              <label
                class="euiFormLabel euiFormRow__label"
                for="type"
              >
                Type
              </label>
            </div>
            <div
              class="euiFormRow__fieldWrapper"
            >
              <div
                id="type"
              >
                <div
                  class="euiRadio euiRadioGroup__item"
                >
                  <input
                    checked=""
                    class="euiRadio__input"
                    id="entry"
                    name="type"
                    type="radio"
                    value=""
                  />
                  <div
                    class="euiRadio__circle"
                  />
                  <label
                    class="euiRadio__label"
                    for="entry"
                  >
                    Link
                  </label>
                </div>
                <div
                  class="euiRadio euiRadioGroup__item"
                >
                  <input
                    class="euiRadio__input"
                    id="menu"
                    name="type"
                    type="radio"
                    value=""
                  />
                  <div
                    class="euiRadio__circle"
                  />
                  <label
                    class="euiRadio__label"
                    for="menu"
                  >
                    Dropdown menu
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--s"
          />
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="name-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="name"
                  >
                    Name
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    class="euiFormControlLayout"
                  >
                    <div
                      class="euiFormControlLayout__childrenWrapper"
                    >
                      <input
                        class="euiFieldText"
                        id="name"
                        name="name"
                        type="text"
                        value="name"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormRow"
                id="dashboard-id-row"
              >
                <div
                  class="euiFormRow__labelWrapper"
                >
                  <label
                    class="euiFormLabel euiFormRow__label"
                    for="dashboard-id"
                  >
                    Dashboard ID or URL
                  </label>
                </div>
                <div
                  class="euiFormRow__fieldWrapper"
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    class="euiComboBox"
                    role="combobox"
                  >
                    <div
                      class="euiFormControlLayout"
                    >
                      <div
                        class="euiFormControlLayout__childrenWrapper"
                      >
                        <div
                          class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                          data-test-subj="comboBoxInput"
                          tabindex="-1"
                        >
                          <span
                            class="euiComboBoxPill euiComboBoxPill--plainText"
                            dashboard_id=""
                          />
                          <div
                            class="euiComboBox__input"
                            style="font-size: 14px; display: inline-block;"
                          >
                            <input
                              aria-controls=""
                              data-test-subj="comboBoxSearchInput"
                              id="dashboard-id"
                              role="textbox"
                              style="box-sizing: content-box; width: 2px;"
                              value=""
                            />
                            <div
                              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                            />
                          </div>
                        </div>
                        <div
                          class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                        >
                          <button
                            aria-label="Clear input"
                            class="euiFormControlLayoutClearButton"
                            data-test-subj="comboBoxClearButton"
                            type="button"
                          >
                            <div
                              class="euiFormControlLayoutClearButton__icon"
                              data-euiicon-type="cross"
                            />
                          </button>
                          <button
                            aria-label="Open list of options"
                            class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                            data-test-subj="comboBoxToggleListButton"
                            type="button"
                          >
                            <div
                              aria-hidden="true"
                              class="euiFormControlLayoutCustomIcon__icon"
                              data-euiicon-type="arrowDown"
                            />
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="euiModalFooter"
    >
      <button
        class="euiButtonEmpty euiButtonEmpty--primary"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            Cancel
          </span>
        </span>
      </button>
      <button
        class="euiButton euiButton--primary euiButton--fill"
        type="submit"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Save
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
